Restricting who can overwrite a state event (#3757)
matrix-org/matrix-spec-proposals: proposals/3757-restricting-who-can-overwrite-a-state-event.md

> # MSC3757: Restricting who can overwrite a state event.

Reviewer: @gleachkr says:

> Here's a small worry - I was hoping that this might resolve one of the [potential issues](https://github.com/opentower/matrix-doc/blob/main/proposals/3574-resource-markup.md#potential-issues) listed for MSC3574, basically that anyone can overwrite an `m.space.child` event sent by someone else. But if the `state_key` is used semantically to indicate ownership, then this is going to be incompatible with any state event type where the `state_key` is used semantically to indicate something else (as in the case of `m.space.child`).

Reviewer 2: true.

Reviewer 3: Unless you also included support for the `state_key` of `m.space.child` being in the form of `@user:domain_!roomid`. But I agree that it's a bit clunky.

Reviewer 4: That'd probably work for most practical purposes. Folks with very long, but still valid, usernames might run up against the 255 byte limit for `state_key`. But I guess that's a more general problematic edge-case - certain valid usernames will consume all the available bytes in `the state_key` (similar to [this issue](https://github.com/matrix-org/matrix-spec-proposals/pull/1769#discussion_r614360884) with extensible profiles).

Author: Resolving this conversation because I don't think it requires changes to this MSC: feel free to unresolve if there is more to say here.

> # MSC3757: Restricting who can overwrite a state event.

Reviewer 5: I don't see it mentioned anywhere what characters are allowed for the "rest" bit of the state event key, is that defined somewhere in the spec, or is this defacto undefined?

If undefined, should we choose a character set?

Author: Good question.

Reviewer 6: Do we actually need an additional restriction here? Given the definition of splitting the state key on the first `_` after the first `:`, the "rest" bit could just contain any character allowed in state keys.

Reviewer 7: Just because state_key lets you include kanji and emoji doesn't *necessarily* mean we should allow it here, but possibly.

The main thing is to be explicit about what is allowed, so that implementations that have to parse the thing don't get any nasty surprises.

Author: I don't see any reason for this MSC to change the rules about what is allowed in a state key, and I want it to benefit from any future improvements to that, so I will add a sentence about the allowed characters matching what is allowed in a state key.

events of the same type in a room, but would like to set access control so
that only they can subsequently update the event. An example of this is if a
user wishes to publish multiple live location share beacons as per [MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489)
and [MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672), for instance one per device.  They will typically not want
other users in the room to be able to overwrite the state event,
so we need a mechanism to prevent other peers from doing so.

[MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489) originally proposed that the event type could be made variable,
appending an ID to each separately posted event so that each one could
separately be locked to the same mxid in the state_key.  However, this is
problematic because you can't proactively refer to these event types in the
`events` field of the `m.room.power_levels` event to allow users to post
them - and they also are awkward for some client implementations to
manipulate.

## Proposal

Therefore, we need a different way to state that a given state event may only

> be written by its owner. **We propose that if a state event's state_key *starts with* a matrix ID (followed by an underscore), only the sender with that matrix ID (or higher PL users) can set the state event.**  This is an extension of the current behaviour where state events may be overwritten only if the sender's mxid *exactly equals* the state_key.

Reviewer 5: I think making it an underscore is problematic for discerning where a domain ends and another starts, domains can have a bunch of ASCII symbols, and while some/most prevent `_` from being used, i believe its a common stance, instead of an absolute rule.

I suggest a backslash being used here instead, as its quite impossible to have domains with those, so this'd be `@user:example.com/<whatever>` as a key.

Reviewer: Underscores are not permitted in domain names in Matrix: https://spec.matrix.org/v1.2/appendices/#server-name

Closing this thread. Let's please not bikeshed on which characters to use.

Reviewer 8: > Let's please not bikeshed

Why? What is the MSC process even for if not for finding out the very best way to implement a certain feature?

> be written by its owner. **We propose that if a state event's state_key *starts with* a matrix ID (followed by an underscore), only the sender with that matrix ID (or higher PL users) can set the state event.**  This is an extension of the current behaviour where state events may be overwritten only if the sender's mxid *exactly equals* the state_key.

Reviewer 2: You end up with a race over selecting a state_key (so an attacker could try to clash your state out of existence). Rather than unravelling that race and its impact on state res, we’re extending the current behaviour

We also allow users with higher PL than the original sender to overwrite state 
events even if their mxid doesn't match the event's state_key. This fixes an abuse
vector where an unprivileged user can immutably graffiti the state within a room
by sending state events whose state_key is their matrix ID.

Practically speaking, this means modifying the [authorization rules](https://spec.matrix.org/v1.2/rooms/v9/#authorization-rules) such that rule 8:

> If the event has a `state_key` that starts with an `@` and does not match the `sender`, reject.

becomes:

> If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

For example, to post a live location sharing beacon from [MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672):

```json=
{
    "type": "m.beacon_info",

>     "state_key": "@stefan:matrix.org_assetid1", // Ensures only the sender can update

Reviewer 3: ```suggestion
    "state_key": "@stefan:matrix.org_assetid1", // Ensures only the sender or higher PL users can update
```

By scoping who can set the state_key to be the mxid of the sender, this problem
goes away.

Another option would be to include the `sender` (or a dedicated `state_subkey`)
as the third component of what makes a state event unique. This is, however, a
much bigger change and not justifiable for the goal meant to be reached with
this MSC.

## Security considerations

As this changes auth rules, we should think carefully about whether could
introduce an attack on state resolution. For instance: if a user had higher
PL at some point in the past, will they be able to abuse somehow this to
overwrite the state event, despite not being its owner?

When using a state_key prefix to restrict who can write the event, we have
deliberately chosen an underscore to terminate the mxid prefix, as underscores
are not allowed in domain names.  A pure prefix match will **not** be sufficient,
as `@matthew:matrix.org` will match a state_key of form `@matthew:matrix.org.evil.com:id1`.

This changes auth rules in a backwards incompatible way, which will break any
use cases which assume that higher PL users cannot overwrite state events whose 

> state_key is a different mxid.  This is considered a feature rather than a bug,

Reviewer 3: Quickly mentioning that this change requires a new room version and thus will not affect old events may be worthwhile.

Author: Done, thanks.

fixing an abuse vector where unprivileged users could send arbitrary state events
which could never be overwritten.

## Unstable prefix

> None

Reviewer: To unblock implementation:
```suggestion
While this MSC is not considered stable, implementations should apply the behaviours of this MSC on top of room version 9 as `org.matrix.msc3757`.
```

If the MSC changes in a way where previous implementations wouldn't be compatible, bump the prefix to something like `org.matrix.msc3757.v2`.

Author: Where would this prefix apply?

Reviewer 6: It's the room version used while this is not stable. I had the same question last week which led to https://github.com/matrix-org/matrix-spec/issues/1004. 😅

> # MSC3757: Restricting who can overwrite a state event.

Reviewer 2: The SCT discussed this during/after a retro last week and concluded that MSC3757 is preferable to MSC3760 and MSC3761 because:

 * It's a very minimal extension to the current behaviour, and does what's needed
 * It doesn't involve redefining what a state event is on every server & client in existence (as MSC3760 does)
 * It doesn't involve an entirely new ACL model, which would need to be very carefully analysed and threat-modelled in terms of state res (as MSC3761 does).

MSC3761 can then land in future once matured.

Reviewer 9: I understand the reasoning, but I can't help wonder whether this will eventually be considered to be cruft and overly limiting, since this proposal *is* mostly equivalent to the state sub-keys proposal, just encoded in a clunkier, but more-though-not-completely backward-compatible way.

Some of the same problems with state sub-keys apply (i.e. what if we ever need subsub-keys and [this comment](https://github.com/matrix-org/matrix-spec-proposals/pull/3757#discussion_r835563980)) and some new ones are created in turn (limitations arising due to the fact that we are now mux'ing two separate pieces of information into a single `state_key` string, such as length limits).

Author: I think this conversation does not propose any changes to this MSC, so I'm resolving it. Please unresolve if needed,

> 

Reviewer 10: I much prefer this solution over MSC3760. This doesn't break data structures and tweaks a single conditional on the auth rules.

It is worth considering whether this data model would need to be extended again in the future, as it seems like we are creating a file system hierarchy at this point, where events "under" your user ID are owned by you (but root can still modify them). Common extensions may be:
 - `chown`ing a directory -> rename all state key prefixes which have that user ID.
 - nesting an owner beneath another owner e.g `@alice/receipts/@bob` where only Bob can write to that key.
 - aggregation operations e.g give me an array of state events which have this prefix.
 -

Author: If we start getting into complex permissions I think we are getting into [MSC3761](https://github.com/matrix-org/matrix-doc/pull/3761) territory. I'd prefer this MSC to stay simple and consider more complex needs later, when they are concrete.

## Problem

Currently there are three main ways to limit who can overwrite a state event:

 * If a user's PL is greater than the `m.room.power_levels` `state_default` field
 * If a user's PL is greater than the `m.room.power_levels` `events` field for that event type
 * If a state event has a state key which begins with an `@`, then the sender's mxid must match that state key.

> This is problematic if an unprivileged user needs to publish multiple state

Reviewer 9: What is the role of the word "unprivileged" here? Unprivileged in what sense?

Author: A user with power level 0. Do you think the wording needs clarification?

Reviewer 9: I guess I'm confused because an unprivileged (PL 0) user wouldn't be able to send state events anyway, so I guess this wording is contingent on the implicit assumption that `state_default` (or the specific `events` key for the event's type) is set to 0?

Author: Thank you, well spotted. Fixed in bd4176fd.

events of the same type in a room, but would like to set access control so
that only they can subsequently update the event. An example of this is if a
user wishes to publish multiple live location share beacons as per [MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489)
and [MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672), for instance one per device.  They will typically not want
other users in the room to be able to overwrite the state event,
so we need a mechanism to prevent other peers from doing so.

[MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489) originally proposed that the event type could be made variable,
appending an ID to each separately posted event so that each one could
separately be locked to the same mxid in the state_key.  However, this is
problematic because you can't proactively refer to these event types in the
`events` field of the `m.room.power_levels` event to allow users to post
them - and they also are awkward for some client implementations to
manipulate.

> ## Proposal

Reviewer: My general concern is that we already have JSON available to us, so we should use that. String packing works in areas where we don't have as fine control (voip call candidates, for example), but for something like this we can and should afford fields.

Author: Since the SCT came down on the side of this proposal vs [MSC3760](https://github.com/matrix-org/matrix-spec-proposals/pull/3760) I consider this resolved. Please unresolve if you disagree.

proposes to include a dedicated `state_subkey` as the third component of what
makes a state event unique.

## Security considerations

This change requires a new room version, so will not affect old events.

As this changes auth rules, we should think carefully about whether could
introduce an attack on state resolution. For instance: if a user had higher
PL at some point in the past, will they be able to abuse somehow this to
overwrite the state event, despite not being its owner?

When using a state_key prefix to restrict who can write the event, we have
deliberately chosen an underscore to terminate the mxid prefix, as underscores
are not allowed in domain names.  A pure prefix match will **not** be sufficient,
as `@matthew:matrix.org` will match a state_key of form `@matthew:matrix.org.evil.com:id1`.

This changes auth rules in a backwards incompatible way, which will break any
use cases which assume that higher PL users cannot overwrite state events whose 
state_key is a different mxid.  This is considered a feature rather than a bug,
fixing an abuse vector where unprivileged users could send arbitrary state events
which could never be overwritten.

> 

Reviewer 11: I have been attacked by spambots! help me!

> 

Author: If they have permission to set a state event, can't they already spam unlimited state events into a room using arbitrary state keys?

> 

Reviewer 12: Yeah, I suppose that's true.

> # MSC3757: Restricting who can overwrite a state event.

Reviewer 7: I think this is actually *de*restricting who can overwrite a state event.

```suggestion
# MSC3757: Relaxing the restrictions on who can overwrite a state event.
```

Reviewer 13: That's true for MSC3779, but not as much for this one; the former bypasses PL restrictions for setting state that belongs to the sender (where "belongs to" = the state key starts with the sender's MXID), but this MSC does not.

The restriction proposed by this MSC is to prevent state that belongs to a particular user from being overwritten by other, equally-powerful users.

The only PL restriction that's relaxed by this MSC is for allowing _more_ powerful users to overwrite state that doesn't belong to them, for the sake of having a tool against state graffiti.

Reviewer 7: You're right. Still, I find the title a bit confusing. How about:

```suggestion
# MSC3757: Extending the set of write-protected state keys
```
or something

Reviewer 7: Revisiting this a few months later (sorry), I still find my suggestion clearer.

## Problem

Currently there are three main ways to limit who can overwrite a state event:

 * If a user's PL is greater than the `m.room.power_levels` `state_default` field
 * If a user's PL is greater than the `m.room.power_levels` `events` field for that event type

>  * If a state event has a state key which begins with an `@`, then the sender's mxid must match that state key.

Reviewer 7: I find this a confusing way of phrasing the current restrictions.

I'd say that there are essentially *two* current restrictions, imposed by rules 7 and 8 in the [auth rules](https://spec.matrix.org/v1.11/rooms/v11/#authorization-rules).

In short:

1. The user's PL must not be less than the "required power level" for that event.
2. In addition, if a state event has a state key which begins with an `@`, then the sender's mxid must match that state key.

The spec for [`m.room.power_levels`](https://spec.matrix.org/v1.11/client-server-api/#mroompower_levels) defines required power level, but in summary, it is determined by the setting in `m.room.power_levels` `events` field for that event type if present; otherwise, it is determined by the `state_default` field.

Reviewer 13: 9f4f31a8

This is problematic if a user needs to publish multiple state
events of the same type in a room, but would like to set access control so
that only they can subsequently update the event. An example of this is if a
user wishes to publish multiple live location share beacons as per [MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489)
and [MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672), for instance one per device.  They will typically not want
other users in the room to be able to overwrite the state event,

> so we need a mechanism to prevent other peers from doing so.

Reviewer 14: I'll let @AndrewFerr talk more about his use case, but broadly I think its for per-device call state within a room.

[MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489) originally proposed that the event type could be made variable,
appending an ID to each separately posted event so that each one could
separately be locked to the same mxid in the state_key.  However, this is
problematic because you can't proactively refer to these event types in the
`events` field of the `m.room.power_levels` event to allow users to post
them - and they also are awkward for some client implementations to

> manipulate.

Reviewer 13: ae17437b

## Proposal

Therefore, we need a different way to state that a given state event may only

> be written by its owner. **We propose that if a state event's state_key *starts with* a matrix ID (followed by an underscore), only the sender with that matrix ID (or higher PL users) can set the state event.**  This is an extension of the current behaviour where state events may be overwritten only if the sender's mxid *exactly equals* the state_key.

Reviewer 13: Yes, the underscore is just a delimiter.

We also allow users with higher PL than the original sender to overwrite state
events even if their mxid doesn't match the event's state_key. This fixes an abuse
vector where a user can immutably graffiti the state within a room
by sending state events whose state_key is their matrix ID.

Practically speaking, this means modifying the [authorization rules](https://spec.matrix.org/v1.2/rooms/v9/#authorization-rules) such that rule 8:

> If the event has a `state_key` that starts with an `@` and does not match the `sender`, reject.

becomes:

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Author: > To overwrite another user's state event, must your power level exceed the event type's PL, or the PL of the user who owns the event?

I was thinking of the event type's power level i.e. we're not changing anything here except expanding the existing exception where a user can create a state event beyond their power level if they use a specific state key.

Could the wording be improved to clarify this?

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 13: > a user can create a state event beyond their power level if they use a specific state key

This isn't mentioned anywhere, at least not explicitly.  If the only intended change to the auth rules is the change to step 8, then the preceding step still says to reject if the sender's PL is less than the event type’s required power level.

For the exception to hold, step 7 could simply be removed, meaning the PL restriction would be applied by the proposed change to rule 8.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 13: Actually, no -- that would remove PL restrictions for non-state events.

How about replacing steps 7 and 8 with this:

- If the event type’s _required power level_ is greater than the `sender`’s power level, and the event does not have a `state_key` matching the `sender` optionally followed by a substring prefixed with `_`, reject.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Author: Hang on, I was getting confused with https://github.com/matrix-org/matrix-spec-proposals/pull/3779 - sorry for the confusion!

I think this MSC _restricts_ who can overwrite certain state events, and I agree that the wording as it is now doesn't make sense, because it actually allows everyone to overwrite, because we won't reach rule 8 without first passing rule 7 which already checks power level.

I think I just prefer https://github.com/matrix-org/matrix-spec-proposals/pull/3779 over this one, and if I understand right, we don't need both.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 13: Circling back on this: since it's not this MSC that bypasses event type power levels, I believe [my proposed rewording to the auth rule](https://github.com/matrix-org/matrix-spec-proposals/pull/3757#discussion_r1627865001) is correct.

Also, with that rewording, perhaps this rule would do the same job as some of the auth rules related to the PLs of the sender & target users of `m.room.member` events (like the sender being able to kick/ban only target users of a lower PL than them), in which case they could be removed to simplify the auth rules somewhat.  (The rules for invite & join would need to stay the same though, lest a powerful user could force anyone to join).

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 7: I don't think this makes sense as currently written. We can only get to rule 8 if we've already passed rule 7, which means that the sender's power level is not less than the required power level.

Assuming you mean "not less than" where you've written "greater than", then this whole rule becomes a no-op.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 7: Last time we changed the auth rules, there was a lot of discussion around "let's never do this again without test vectors for the auth rules" (cf https://github.com/matrix-org/matrix-spec/issues/1710, https://github.com/matrix-org/matrix-spec/issues/1642).

The apparent confusion over what the current auth rules mean, and what they should say after the change, makes me think that we should start some progress in that direction.

Reviewer 13: The [Complement tests](https://github.com/matrix-org/complement/pull/733) should be a step in the right direction.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 7: /me rereads comment thread properly. @AndrewFerr's [proposal](https://github.com/matrix-org/matrix-spec-proposals/pull/3757#discussion_r1627865001) sounds correct.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 13: I've reworded the auth rule in 5e95ff32. I also added a condition to reject state keys beginning with `@` that are not valid MXIDs, which prevents the power level comparison from looking up the PL of something that is not a user.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 13: No, because
1. The only significant `:` for this parsing is the separator between an MXID's localpart and its domain, and is thus unaffected by `:`s appearing in IPv6 domains
2. `_` is not an allowed character for IPv6 domains, at least not according to [the spec](https://spec.matrix.org/v1.11/appendices/#server-name)

The latter point is worth mentioning in the MSC.

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 13: dd9b33e1

> > If the event has a `state_key` that starts with an `@`, and the substring before the first `_` that follows the first `:` (or end of string) does not match the `sender`, reject - unless the sender's powerlevel is greater than the event type's *required power level*.

Reviewer 13: IMO the auth rules are clearer now, so marking this as resolved.

No additional restrictions are made about the content of the `state_key`, so any characters that follow the `sender` + `_` part are only required to be valid for use in a `state_key`.

For example, to post a live location sharing beacon from [MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672):

```json=
{
    "type": "m.beacon_info",
    "state_key": "@stefan:matrix.org_assetid1", // Ensures only the sender or higher PL users can update
    "content": {
        "m.beacon_info": {
            "description": "Stefan's live location",
            "timeout": 600000,
            "live": true
        },
        "m.ts": 1436829458432,
        "m.asset": {
            "type": "m.self"
        }
    }
}
```

> Since `:` is not permitted in the localpart and `_` is not permitted in the domain part of an mxid (see [Historical User IDs](https://spec.matrix.org/v1.2/appendices/#historical-user-ids)), it is not possible to craft an mxid that matches the beginning of a state key constructed for another user's mxid, so state keys restricted to one owner can never be overwritten by another user.

Reviewer 15: While the spec specifies a grammar for server names, are we SURE that servers verify that grammar? While officially underscores are invalid in domain names and you can't order a TLS certificate for it nowadays, such domains still exist in the wild and wildcard certificates even allow using TLS with them. For example: https://my_sarisari_store.typepad.com/

I think relying on underscores as a separator is a rather scary trap and I wouldn't bet on all currently developed Matrix servers rejecting such server names.

Reviewer 13: ac245102 adds this to the "Potential issues" section.

Reviewer 13: How about using `@` as the separator?  Since the separator will never be the first character in the state key, it shouldn't be possible to confuse it with the user ID sigil.

Reviewer 13: `|` could also be a good separator.

## Potential issues

> None yet.

Reviewer 16: This is going to be incompatible with some Matrix IDs:

> `state_key` MUST NOT exceed 255 bytes. [source](https://spec.matrix.org/unstable/client-server-api/#size-limits)

> The length of a user ID, including the `@` sigil and the domain, MUST NOT exceed 255 bytes. [source](https://spec.matrix.org/unstable/appendices/#user-identifiers)

If you have a user at the maximum length (or close to it) then there's no room for a suffix.

Reviewer 15: That's also my main concern. There is no guarantee on how large the suffix can be, so users will get varying experiences depending on how long their mxid is. This MSC should specify new sizes for mxids, state keys and as a result the suffix, that can be appended in a predictable manner. Alternatively, since this requires a new room version anyway and would be a breaking change for clients, https://github.com/matrix-org/matrix-spec-proposals/pull/3760 would probably have the same impact on implementations.

Reviewer 16: > This MSC creating further implicit limits on participation for "silly" user IDs is a +1 from me.

I'm having trouble parsing this sentence. You think that it isn't a problem to have parts of the protocol incompatible with each other?

Having MXIDs "usually" banned in those situations is not a good enough reason IMO. I feel strongly about that more internal inconsistency is bad for the project.

Reviewer: The inconsistency doesn't bother me because I would favour shortening the allowable length anyways. I see this MSC as a very small stepping stone in that direction (we'd resolve the inconsistency in favour of shorter user IDs, in my opinion/dream).

Reviewer 13: eb0eed6d adds this to the "Potential issues" section.

Also, I came across an existing issue with MXID size limits: a homeserver can have a server name that's 253 characters long (which is permitted by both [RFC 1035 §2.3.4](https://www.rfc-editor.org/rfc/rfc1035#section-2.3.4) and the lack of a size limit on server names), which would force all MXIDs on that homeserver to be over 255 characters long (253 for the server name + 1 for the `@` sigil + 1 for the `:` separator).  So perhaps a general reworking of MXID & state key size limits isn't out of the question.

Reviewer 16: Yes, that's a bit of an odd issue, but I think this current MSC would exacerbate the issue quite a lot with existing homeservers.

Reviewer 13: 590ff96c & d9b149d8 address this issue.

> None yet.

Reviewer 13: > * (a) ... This MSC would change that, by allowing any user to (say) create millions of `m.beacon_info` events.

This is what MSC3779 proposes, not this one.

To clarify, the effective auth rule changes proposed by this MSC are to:
- grant write access on state events with state keys of the form `<sender>_<anySuffix>` to all users
  - N.B. `<sender>` is the MXID of the user sending the state event
- grant write access on state events with state keys of the form `<mxid>` & `<mxid>_<anySuffix>` to users whose PL is greater than that of `<mxid>`'s PL
  - N.B. `<mxid>` is any MXID

The event-sending restrictions imposed by `m.room.power_levels` remain in effect.

What this MSC _does_ allow is creating lots of state events with different state key suffixes -- but that would be limited to room moderators, just like how creating lots of state events with different types currently is.

> * (b) there is currently no _specced_ usecase that requires multiple state events per user, and introducing one feels like a step-change.

As a start, MSC4143 makes an explicit mention to this one (namely, in [the section regarding MatrixRTC room state](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#the-matrixrtc-room-state)), and is what has driven interest in this MSC.

> None yet.

Reviewer 7: >> This MSC would change that, by allowing any user to (say) create millions of m.beacon_info events.

You're right; that is not a change proposed by this MSC, so my argument was wrong.

> To clarify, the effective auth rule changes proposed by this MSC are to:
> 
>  * grant write access on state events with state keys of the form `<sender>_<anySuffix>` to all users

... but only if the user has a high enough PL to send that event, right?

I still have reservations about any change which steers us towards having more state events in a room -- and expanding the set of state keys that are restricted to specific users sounds like it is going to do just that.

But maybe that's an argument for another MSC.

> ## Alternatives

Reviewer: String-based matching still feels like a bandaid solution for a larger problem. The problem statement is reasonable (beacons, calls, and integrations all want an ability to narrow power levels), however there are non-state events where it'd be worth having a similar access mechanism (namely for edits).

At the point where we're considering string-packing, a top level field on the event works just as well and is more extensible, capable, and presentable than a packed string. It can be populated with query parameters to the `PUT /state` or `PUT /send` calls, and servers are easily able to retrieve information from the event. Further, since we're modifying the auth rules to accomplish this, there is no material difference to the implementation effort of the proposal itself.

Reviewer 17: I am wondering if a top level field implies that state_key and access mechanism are independent.

If this is what is proposed I have the concern that this would allow blocking of state keys by not allowing to write to them anymore. (I think I saw this being brought up somewhere else as well. cannot find it right now)

But the concept of merging key + access level makes a lot of sense to me since it would behave more like a namespace than an add-on permission system. And I think the idea of namespaces (kind of like inverse domains where each user owns the domain starting with their username) is a very good fit for a key value based state like the room state.

Author: If I understand it right, this proposal _adds_ new ways a user can create a state event.
```suggestion
## Alternatives
```

The user can create a state event if:

* The power levels permit it (as before), OR
* The state key = their mxid (as before), OR
* The state key has the new format, starting with "mxid_"

It also fixes an existing spec bug where users with power based on power levels were not allowed to modify state where the state_key was someone else's mxid.

So it uses special keys to add more ways you can be allowed to modify state, extending the existing special key.

Author: Nope, the above was wrong - I was mixing this up with https://github.com/matrix-org/matrix-spec-proposals/pull/3779 which I think replaces this, and which I prefer.

Reviewer 13: 63955d72

proposes to include a dedicated `state_subkey` as the third component of what
makes a state event unique.

## Security considerations

This change requires a new room version, so will not affect old events.

As this changes auth rules, we should think carefully about whether could
introduce an attack on state resolution. For instance: if a user had higher
PL at some point in the past, will they be able to abuse somehow this to
overwrite the state event, despite not being its owner?

When using a state_key prefix to restrict who can write the event, we have
deliberately chosen an underscore to terminate the mxid prefix, as underscores
are not allowed in domain names.  A pure prefix match will **not** be sufficient,
as `@matthew:matrix.org` will match a state_key of form `@matthew:matrix.org.evil.com:id1`.

This changes auth rules in a backwards incompatible way, which will break any
use cases which assume that higher PL users cannot overwrite state events whose 
state_key is a different mxid.  This is considered a feature rather than a bug,
fixing an abuse vector where users could send arbitrary state events
which could never be overwritten.

## Unstable prefix

> While this MSC is not considered stable, implementations should apply the behaviours of this MSC on top of room version 9 as `org.matrix.msc3757`.

Reviewer 13: f962bf34 changes this to say to apply the MSC onto version >=10.

events of the same type in a room, but would like to set access control so
that only they can subsequently update the event. An example of this is if a
user wishes to publish multiple live location share beacons as per [MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489)
and [MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672), for instance one per device.  They will typically not want
other users in the room to be able to overwrite the state event,
so we need a mechanism to prevent other peers from doing so.

[MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489) originally proposed that the event type could be made variable,
appending an ID to each separately posted event so that each one could
separately be locked to the same mxid in the `state_key`.  However, this is
problematic because you can't proactively refer to these event types in the
`events` field of the `m.room.power_levels` event to allow users to post
them - and they also are awkward for some client implementations to
manipulate.

## Proposal

Therefore, we need a different way to state that a given state event may only

> be written by its owner. **We propose that if a state event's `state_key` *starts with* a matrix ID (followed by an underscore), only the sender with that matrix ID (or higher PL users) can set the state event.**  This is an extension of the current behaviour where state events may be overwritten only if the sender's mxid *exactly equals* the `state_key`.

Reviewer 10: Is it valid to have an empty string as the suffix? E.g `@kegan:matrix.org_`? What about having arbitrary unicode characters? Null bytes? **Please** let's be sensible and force sensible restrictions on this new user-defined variable, lest we end up in another hell of poor validation causing problems for clients/servers.

I would suggest a very strong restriction of `[a-z0-9]` unless there are good reasons to allow other characters (there almost certainly are not).

Reviewer 13: Allowing an empty suffix probably won't hurt.  I can imagine cases of the suffix being a property that may take on an empty/nullish value.

For a device ID to be usable as a suffix, the suffix charset must include all characters that can be used as a device ID. Unfortunately, there's no specced device ID charset that I can find, and in practice it is quite broad:
- Device IDs generated by QR code logins (at least on Element X Android) can contain `/` and `+`.
- [`/_matrix/client/v3/login`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3login) allows setting a custom device ID of any string, and the Synapse implementation allows all sorts of characters (and lets me set a device ID of `a b ?!@#$%^&*()-=_+/`).

So if we want device ID suffixes now, maybe defining a suffix charset is premature, unless this MSC also defines a stricter device ID charset.  But that raises the issue of what to with existing device IDs that don't follow the charset...

Reviewer 13: That's a good idea. I was worried that the input size of device IDs would be too large to safely hash, but the point is to avoid hash _collisions_, the chance of which should still be small.

Then, this MSC can do away with trying to define a suffix length / charset that can fit a raw device ID.

I will still propose a few non-word characters to be in the suffix charset, because it may be handy to have a suffix containing multiple properties that are easy to separate with a non-word character (eg. `@user:hs_prop1_prop2_prop3`).  The spec already uses a charset of `[0-9a-zA-Z.=_-]` for email login secrets/tokens, and would be a decent charset to use here as well.

Reviewer 13: On second thought, a handy consequence of _not_ restricting the suffix charset is the ability to prefix any existing state key with a user's MXID to scope its ownership to that user.  With a restricted suffix charset, there may be some state keys that would be invalid as a suffix.

IMO as long as the charset of state keys in general is left unrestricted, there's little benefit in restricting the suffix charset, unless doing so is a step towards restricting the general state key charset.

Reviewer 18: I agree that there's no point in restricting these specific state keys when all other state keys are still arbitrary strings. To take advantage of this MSC, you must give users permission to send state events of certain types. That means users will be able to send unprefixed state keys too, which will not have any character set restrictions.

Restricting state keys in general might be a good idea to do in combination with MSC2828

Reviewer 10: I don't agree with the attitude of "we don't have good validation, so let's not add validation". There's zero reason to punt this down the line to another MSC. It's absolutely trivial to expand the character set or length limits in another MSC. It's very difficult to restrict it once there are client implementations in the wild relying on there being no restrictions. Validation is important to reduce the attack surface of any newly added features. The suffix string will be stored in new places where the state key is not (e.g DBs will likely either have this as a column, or indexed in such a way to allow efficient ordered lookups), which means there will be new code written to read this input. Validate the input, please.

Reviewer 13: If the intent of a restricted suffix charset is to ease parsing out the user ID prefix, then the usefulness of the restriction might be limited by how the prefix must be able to contain any printable ASCII character, due to having to support [historical user IDs](https://spec.matrix.org/v1.11/appendices/#historical-user-ids).  Also, since the parsing is concerned more with the user ID prefix than with the suffix, locking down the suffix charset might not impact much.

Besides, the "suffix string" isn't really meant to be a new kind of syntax for state keys, but is just a result of the semantic of user-scoped state keys via string packing.  It's perhaps better to think of a scoped state key not as a user ID + a suffix, but as an ordinary state key prefixed by a user ID; or rather, that this MSC proposes allowing state keys to optionally be prefixed by a user ID.  Being prefixed doesn't change the nature of the content of the rest of the state key, but applying different restrictions to prefixed & unprefixed state keys would imply that it does.

Though if restricting the charset is non-negotiable, then maybe a compromise is to apply a broad charset, like all printable ASCII characters.

events even if their mxid doesn't match the event's `state_key`. This fixes an abuse
vector where a user can immutably graffiti the state within a room
by sending state events whose `state_key` is their matrix ID.

Practically speaking, this means modifying the [authorization rules](https://spec.matrix.org/v1.2/rooms/v9/#authorization-rules) such that rule 8:

> 8. If the event has a `state_key` that starts with an `@` and does not match the `sender`, reject.

becomes:

> 8. If the event has a `state_key` that starts with an `@`:
>    1. If the prefix of the `state_key` before the first `_` that follows the first `:` (or end of string) is a valid user ID:
>       1. If that user ID does not match the `sender`, and the `sender`'s power level is not greater than that of the user denoted by the ID, reject.
>    2. Otherwise, reject.

No additional restrictions are made about the content of the `state_key`, so any characters that follow the `sender` + `_` part are only required to be valid for use in a `state_key`.

For example, to post a live location sharing beacon from [MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672):

```json=
{
    "type": "m.beacon_info",

>     "state_key": "@stefan:matrix.org_assetid1", // Ensures only the sender or higher PL users can update

Reviewer 10: It's not clear what `assetid1` ultimately means semantically. The use case is for device level updates of beacon data, so why isn't this just `$USERID_$DEVICEID` and we can do away with yet more user-defined input?

Reviewer 13: 486b0cd8

Or, did you mean that device IDs should be the only possible suffix for all event types?

    "content": {
        "m.beacon_info": {
            "description": "Stefan's live location",
            "timeout": 600000,
            "live": true
        },
        "m.ts": 1436829458432,
        "m.asset": {
            "type": "m.self"
        }
    }
}
```

Since `:` is not permitted in the localpart and `_` is not permitted in the domain part of an mxid (see [Historical User IDs](https://spec.matrix.org/v1.2/appendices/#historical-user-ids)), it is not possible to craft an mxid that matches the beginning of a state key constructed for another user's mxid, so state keys restricted to one owner can never be overwritten by another user.

## Potential issues

### Incompatibility with long user IDs

As the spec enforces [a size limit of 255 bytes for both MXIDs and state keys](https://spec.matrix.org/unstable/client-server-api/#size-limits),
the set of available MXID-prefixed state keys is smaller for long MXIDs than for short ones,
with the worst case of none being available for MXIDs equal to the size limit.
Thus, long MXIDs are restricted from being used as state key prefixes to designate state ownership.

> This issue could be solved by increasing the size limit for state keys.

Reviewer 13: For a device ID to be usable as a suffix, the max suffix length must be large enough to contain one.  Unfortunately, device IDs don't seem to have a specced size limit, and Synapse applies a limit of a hefty 512 characters.

The simplest solution is to propose a `max_suffix_len` of 512, and perhaps to define a max device ID length as well.

> This issue could be solved by increasing the size limit for state keys.

Reviewer 13: If the practice of building suffixes from hashes is encouraged, then a shorter max suffix length should be sufficient.

590ff96c proposes a max suffix length of 255 bytes, which is just enough room to allow using any currently-valid state key as a suffix.

> This issue could be solved by increasing the size limit for state keys.

Reviewer 13: It may also be helpful to keep the 255-byte limit for state keys without a user ID prefix, so that any unprefixed key could also be used with a prefix without surpassing the new 511-byte size limit for the whole key.

It would also make it clear that the purpose of the size increase is not to allow for longer state keys in general, but is just a means of having enough room for a full-length user ID prefix.

> This issue could be solved by increasing the size limit for state keys.

Reviewer 13: d9b149d8

`events` field of the `m.room.power_levels` event to allow users to post
them - and they also are awkward for some client implementations to
manipulate.

An earlier draft of this MSC proposed putting a flag on the contents of the
event (outside of the E2EE payload) called `m.peer_unwritable: true` to indicate
if other users were prohibited from overwriting the event or not.  However, this
unravelled when it became clear that there wasn't a good value for the `state_key`,
which needs to be unique and not subject to races from other malicious users.
By scoping who can set the `state_key` to be the user ID of the sender, this problem
goes away.

[MSC3760](https://github.com/matrix-org/matrix-spec-proposals/pull/3760)
proposes to include a dedicated `state_subkey` as the third component of what
makes a state event unique.
As an extension to this idea, a comment in [the discussion of this MSC](
https://github.com/matrix-org/matrix-spec-proposals/pull/3757#issuecomment-2099010555)
proposes allowing `state_key` to be an array of strings.
Either proposal allows for effectively including an owning user ID in a state key without having to
string-pack the user ID with another string.

> However, either proposal would alter the nature of state events and state resolution.

Reviewer 16: Doesn't this proposal also alter the nature of state events & state res? This isn't really enough evidence to discard this alternative IMO.

Reviewer 13: IMO it doesn't, because the most significant changes pertain only to the auth rules of who can set a state event.  The mechanism of identifying state events by type + state key is unchanged, and the proposed changes to state keys (like the size limit increase & how to allow user ID prefixes) are AFAICT backwards-compatible with existing state events & state res.

Reviewer 16: It says "either proposal" -- the implication to me is that the complexity added be either is roughly the same. From your comments it sounds like you disagree.

Reviewer 13: This meant to refer to both ideas of multi-component state fields (either a state subkey or a variable-length state key array). I'll reword this to be more clear.

Reviewer 13: 5855a7fa

 * Only users with a power level greater than or equal to the "required power level" for a state
   event type may send state events of that type.
 * State events with a `state_key` that equals a user ID may be overwritten only by the user whose
   ID matches the state key.

With these restrictions, only a single piece of state for any state event type may have its write
access limited to a particular user (the state event whose `state_key` is set to the ID of the user
who has write access to it).

This is problematic if a user needs to publish multiple state
events of the same type in a room, but would like to set access control so
that only they can subsequently update the event. An example of this is if a
user wishes to participate in VoIP call as per [MatrixRTC (MSC4143)](
https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md#the-matrixrtc-room-state),
by sending a state event for each of their devices that is to participate in the call.
They will typically not want other users in the room to be able to overwrite those state events,

> so there ought to be a mechanism to prevent other users from doing so.

Reviewer 9: To be honest, it's not clear to me that MatrixRTC call events *should* be state events. Call state is very ephemeral information, while room state is generally meant for longer-lived state, like room configuration.

Areas where state did end up being used for quicker evolving state have are precisely those that showcased many problems with such an approach: the need to retain stricter ownership over a subset of state events; the need to allow less privileged users to send state events, which brings about state spam concerns; the large volume of events calling for a mechanism to delete state, which turns out to be an almost impossible feat. Not to mention that state events in today's Matrix are unencrypted. It all screams bad fit to me.

I realise this is coming a bit late to the party given when MatrixRTC implementations are already running in the wild. But given that this is used as the sole example of a requirement for this MSC and there is widespread dissatisfaction with string packing, I think it needs to be raised.

Reviewer 10: I second this. State events is the wrong place for this data, and by putting it in this wrong place we are creating a rod for our own back. String packing is, in the words of other people, a horrible hack. Does it work? Undeniably yes. Do we want to bake this into the protocol permanently? No. This is precisely the job of the SCT to block this kind of thing imo.

Now this does beg the question: what _is_ the right place then? I've banged on internally that there isn't really a right place for this kind of data currently, and have been spending my downtime slowly crafting a proposal for an alternative. Effectively, all we want is:
 - Custom EDUs which can be sent by clients 
 - persistent EDUs which guarantee delivery of the latest value sent on a per-room, per-user basis.

Thus sending a state event only you can modify becomes a persistent EDU (quite an oxymoron given the E stands for Ephemeral!) with whatever custom type/content you want.

The main driver for this MSC afaict is MatrixRTC so we can even drop the "custom" part for now and only have fixed known EDUs for now to reduce the scope and risk of the changes. 

I'd be much happier to thumbs up a persistent EDU proposal than shoehorn in owned state events.

Reviewer 10: > I'd be much happier to thumbs up a persistent EDU proposal than shoehorn in owned state events.

https://github.com/matrix-org/matrix-spec-proposals/pull/4354 is this.

Reviewer 19: Related conversation in the [*Improving the security of the federation protocol*](https://www.youtube.com/watch?v=m_2ZBs_4RbE&t=2829) Matrix Conference 2025 talk by @kegsay. Some general notes:

 - State is for when you need conflict resolution
    - For setting the `m.room.name`, there can only be one `m.room.name` so you need conflict resolution
    - We should factor out things that don't need to be in state. Stay in your lane traffic like profile information (and MatrixRTC) should not be in state because it doesn't need conflict resolution
 - Would be good to get guidelines in the spec for when to use state vs the newly proposed [sticky events](https://github.com/matrix-org/matrix-spec-proposals/pull/4354) vs to-device messages and EDU's to prevent people from abusing them in the future.

## Proposal

In a future room version,
**if a state event's `state_key` *starts with* a user ID followed by an underscore, only the user
with that ID or users with a higher power level then them may overwrite that state event.**
This is an extension of the current behaviour where state events may be overwritten only by users
whose ID *exactly equals* the `state_key`.

As the spec currently enforces [a size limit of 255 bytes for both user IDs and state keys](
https://spec.matrix.org/unstable/client-server-api/#size-limits),
the size limit on state keys is increased to **511 bytes** to allow prefixing any currently-valid
state key with a maximum-length user ID (and a separator character).
The size of a state key suffix after a leading user ID and the separator character is limited to
**255 bytes** so that any such suffix may follow any user ID without the complete state key
ever surpassing the total state key size limit.
Similarly, the size of a state key without a leading user ID is limited to **255 bytes** so that any

> state key without a leading user ID may be given one without ever surpassing the total size limit.

Reviewer 20: This part can most charitably be described as an awful hack. There should never be a reason to encode multiple independent values into a single string field *inside a structured data format like JSON*, let alone then enforcing strict length limits on different parts. No, this is terrible, the section "Multi-component state keys" describes what a real solution would have to look like.

Note that the size limit of 256 bytes after a leading user ID includes the separating `_`.

No additional restrictions are made about the content of the `state_key`, so any characters that
follow the `sender` + `_` part are only required to be valid for use in a `state_key`.

For example, to post a live location sharing beacon from
[MSC3672](https://github.com/matrix-org/matrix-spec-proposals/pull/3672)
for one of a user's devices:

```json=
{
    "type": "m.beacon_info",
    "state_key": "@stefan:matrix.org_{deviceid1}", // Ensures only the sender or higher PL users can update
    "content": {
        "m.beacon_info": {
            "description": "Stefan's live location",
            "timeout": 600000,
            "live": true
        },
        "m.ts": 1436829458432,
        "m.asset": {
            "type": "m.self"
        }
    }
}
```

## Potential issues

### Incompatibility with domain names containing underscores

Although both [the spec](https://spec.matrix.org/unstable/appendices/#server-name)
and [RFC 1035 §2.3.1](https://www.rfc-editor.org/rfc/rfc1035#section-2.3.1)
forbid the presence of underscores in domain names,

> there noneless exist resolvable domain names that contain underscores.

Reviewer 19: ```suggestion
there nonetheless exist resolvable domain names that contain underscores.
```

so there ought to be a mechanism to prevent other users from doing so.

## Proposal

In a future room version,
**if a state event's `state_key` *starts with* a user ID followed by an underscore, only the user
with that ID or users with a higher power level then them may overwrite that state event.**
This is an extension of the current behaviour where state events may be overwritten only by users
whose ID *exactly equals* the `state_key`.

As the spec currently enforces [a size limit of 255 bytes for both user IDs and state keys](
https://spec.matrix.org/unstable/client-server-api/#size-limits),
the size limit on state keys is increased to **511 bytes** to allow prefixing any currently-valid
state key with a maximum-length user ID (and a separator character).
The size of a state key suffix after a leading user ID and the separator character is limited to
**255 bytes** so that any such suffix may follow any user ID without the complete state key
ever surpassing the total state key size limit.
Similarly, the size of a state key without a leading user ID is limited to **255 bytes** so that any

> state key without a leading user ID may be given one without ever surpassing the total size limit.

Reviewer 7: Is this conversion between prefixed and unprefixed state keys a real requirement? It seems to lead to a lot of complexity.

If we're doing this, I'd be inclined to just increase max state key length to (say) 512 bytes, and be done with it.

Users with a higher power level than a state event's original sender may overwrite the event
despite their user ID not matching the one in event's `state_key`. This fixes an abuse
vector where a user can immutably graffiti the state within a room

> by sending state events whose `state_key` is their user ID.

Reviewer 7: I'd say there's a strong argument for making this a separate MSC, rather than trying to fix two things at once.
